src/astarte/handler.rs

```diff
@@ -288,24 +288,98 @@ impl AstarteSubscriber for DevicePublisher {
-                if self.client.remove_interface(interface).await? {
-                    removed.push(interface.to_string());
-                }
-            }
-
-            removed
+            self.client.remove_interfaces(to_remove).await?
         };
 
         subscribers.remove(id);
 
         Ok(removed)
     }
+
+    /// Extend the existing Node interfaces and its introspection.
+    async fn extend_interfaces(
+        &self,
+        node_id: &Uuid,
+        to_add: HashMap<String, Interface>,
+    ) -> Result<Vec<Interface>, AstarteMessageHubError> {
+        let mut rw_subscribers = self.subscribers.write().await;
+
+        let Some(subscriber) = rw_subscribers.get_mut(node_id) else {
+            error!("invalid node id {node_id}");
+            return Err(AstarteMessageHubError::NodeId(*node_id));
+        };
+
+        trace!(
+            "extending introspection with the following interfaces: {:?}",
+            to_add.keys().collect_vec()
+        );
+        let added_names = self
+            .client
+            .extend_interfaces_vec(to_add.values().cloned().collect())
+            .await?;
+        info!("Node interfaces extended");
+
+        let added_interfaces = added_names
+            .iter()
+            .filter_map(|i| to_add.get(i).cloned())
+            .collect();
+
+        subscriber.introspection.extend(to_add.into_keys());
+
+        Ok(added_interfaces)
+    }
+
+    /// Remove one or more Node interfaces from its introspection.
+    async fn remove_interfaces(
+        &self,
+        node_id: &Uuid,
+        to_remove: HashSet<String>,
+    ) -> Result<Vec<String>, AstarteMessageHubError> {
+        let mut rw_subscribers = self.subscribers.write().await;
+
+        if !rw_subscribers.contains_key(node_id) {
+            error!("invalid node id {node_id}");
+            return Err(AstarteMessageHubError::NodeId(*node_id));
+        };
+
+        // check that each of the interfaces to remove is not present in other nodes:
+        // - if at least another node uses that interface, the interface should not be removed from the introspection but only from that node (in the subscriber map)
+        // - if none of the other nodes use that interface, it can also be removed from the introspection
+        let to_remove_from_intr = to_remove
+            .iter()
+            .filter(|&iface| {
+                !rw_subscribers
+                    .iter()
+                    // Check if any other subscriber use the same interface
+                    .any(|(s_id, s_node)| s_id != node_id && s_node.introspection.contains(iface))
+            })
+            .cloned()
+            .collect_vec();
+
+        trace!("removing the following interfaces from the introspection: {to_remove_from_intr:?}");
+        let removed_interfaces = self
+            .client
+            .remove_interfaces_vec(to_remove_from_intr)
+            .await?;
+        info!("interfaces removed");
```

Comment:
change to debug or trace
```suggestion
        trace!("interfaces removed");
```